spec/event_builder_spec.rb

```diff
@@ -364,23 +357,71 @@
 
   # Create impression event with bucketing ID
   it 'should create valid Event when create_impression_event is called with Bucketing ID attribute' do
-    @expected_impression_params[:visitors][0][:attributes] = [{
-      entity_id: '111094',
-      key: 'browser_type',
+    @expected_impression_params[:visitors][0][:attributes].unshift({
+                                                                     entity_id: '111094',
+                                                                     key: 'browser_type',
+                                                                     type: 'custom',
+                                                                     value: 'firefox'
+                                                                   },       entity_id: Optimizely::Helpers::Constants::CONTROL_ATTRIBUTES['BUCKETING_ID'],
+                                                                            key: Optimizely::Helpers::Constants::CONTROL_ATTRIBUTES['BUCKETING_ID'],
+                                                                            type: 'custom',
+                                                                            value: 'variation')
+
+    user_attributes = {
+      'browser_type' => 'firefox',
+      Optimizely::Helpers::Constants::CONTROL_ATTRIBUTES['BUCKETING_ID'] => 'variation'
+    }
+    experiment = config.get_experiment_from_key('test_experiment')
+    impression_event = @event_builder.create_impression_event(experiment, '111128', 'test_user', user_attributes)
+    expect(impression_event.params).to eq(@expected_impression_params)
+    expect(impression_event.url).to eq(@expected_endpoint)
+    expect(impression_event.http_verb).to eq(:post)
+  end
+
+  it 'should create valid Event with user agent when bot_filtering is enabled' do
+    # Test that create_impression_event creates Event object
+    # with right params when user agent attribute is provided and
+    # bot filtering is enabled
+
+    @expected_impression_params[:visitors][0][:attributes].unshift(
+      entity_id: Optimizely::Helpers::Constants::CONTROL_ATTRIBUTES['USER_AGENT'],
+      key: Optimizely::Helpers::Constants::CONTROL_ATTRIBUTES['USER_AGENT'],
       type: 'custom',
-      value: 'firefox'
-    }, {
-      entity_id: OptimizelySpec::RESERVED_ATTRIBUTE_KEY_BUCKETING_ID,
-      key: OptimizelySpec::RESERVED_ATTRIBUTE_KEY_BUCKETING_ID_EVENT_PARAM_KEY,
+      value: 'test'
+    )
+    user_attributes = {
+      Optimizely::Helpers::Constants::CONTROL_ATTRIBUTES['USER_AGENT'] => 'test'
```

Comment:
Use hardcoded $opt_user_agent string in one of the places. Either in user_attributes or in @expected_impression_params